src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts

```diff
@@ -1,100 +1,164 @@
+  private hubConnection: signalR.HubConnection;
 
-  messages: Message[];
+  messages: Message[] = [];
   messagesParameters: MessagesParameters = {
     from: 0,
-    size: 0
+    size: 20
   };
-  isMobileView: boolean;
+  messageControl: FormControl = new FormControl('');
   chatRoom: ChatRoom;
   userRole: Role;
+  userIsProvider: boolean;
+  userName: string;
+  companionName: string;
+  isAdditionalMessageLoading = false;
 
   @Select(ChatState.selectedChatRoomMessages)
   messages$: Observable<Message[]>;
   @Select(ChatState.selectedChatRoom)
   chatRoom$: Observable<ChatRoom>;
+  @Select(SharedUserState.selectedWorkshop)
+  workshop$: Observable<Workshop>;
+  //TODO: Fix parent object formation from two objects
+  @Select(RegistrationState.user)
+  user$: Observable<User>;
+  @Select(RegistrationState.parent)
+  parent$: Observable<Parent>;
   @Select(RegistrationState.role)
   role$: Observable<Role>;
 
   @ViewChild('chat') chatEl: ElementRef;
 
-  @HostListener('window: resize', ['$event.target'])
-  onResize(event: Window): void {
-    this.isMobileView = event.outerWidth < 530;
-  }
-
-  constructor(
-    private store: Store,
-    private route: ActivatedRoute,
-    private location: Location
-  ) {}
+  constructor(private store: Store, private route: ActivatedRoute, private location: Location, private signalRService: SignalRService) {}
 
   ngOnInit(): void {
     this.addNavPath();
-    this.getChatRoom();
     this.getUserRole();
-    this.addListeners();
-    this.onResize(window);
+    this.getChatRoom();
+    this.createHubConnection();
   }
 
   ngAfterViewInit(): void {
-    //TODO: Replace setTimeout
-    setTimeout(() => {
-      this.chatEl.nativeElement.scrollTop =
-        this.chatEl.nativeElement.scrollHeight;
-    }, 100);
+    this.setChatSubscriptions();
+
+    this.chatEl.nativeElement.onscroll = () => {
+      if (this.chatEl.nativeElement.scrollTop < 10) {
+        this.chatEl.nativeElement.scrollTop = 10;
+
+        if (!this.isAdditionalMessageLoading) {
+          this.isAdditionalMessageLoading = true;
+          this.store.dispatch(new GetChatRoomMessages(this.chatRoom.id, this.userRole, { from: this.messages.length, size: 20 }));
+        }
+      }
+    };
+  }
+
+  private getChatRoom(): void {
+    const mode = this.route.snapshot.queryParamMap.get('mode');
+
+    switch (mode) {
+      case 'companion':
+        this.createChatRoom();
+        break;
+      default:
+        const chatRoomId = this.route.snapshot.paramMap.get('id');
+        this.store.dispatch(new GetChatRoomById(chatRoomId, this.userRole));
+    }
+  }
+
+  private createChatRoom(): void {
+    const companionId = this.route.snapshot.paramMap.get('id');
+
+    this.store.dispatch(new GetWorkshopById(companionId));
+
+    combineLatest([this.workshop$, this.user$, this.parent$])
+      .pipe(
+        filter(([workshop, user, parent]) => !!workshop && !!user && !!parent),
```

Comment:
filter(([workshop, user, parent]) => !!workshop && !!user && !!parent),
```suggestion
        filter(([workshop, user, parent]) => !!(workshop && user && parent),
```